crates/net/src/rlpx/ecies.rs

```diff
@@ -0,0 +1,386 @@
+    pub ephemeral_key: SecretKey,
+    pub handshake_data: Option<HandshakeData>,
+}
+
+impl RLPxConnection {
+    pub fn new(nonce: H256, ephemeral_key: SecretKey) -> Self {
+        Self {
+            nonce,
+            ephemeral_key,
+            handshake_data: None,
+        }
+    }
+
+    pub fn random() -> Self {
+        let mut rng = thread_rng();
+        Self::new(H256::random_using(&mut rng), SecretKey::random(&mut rng))
+    }
+
+    pub fn is_connected(&self) -> bool {
+        self.handshake_data.is_some()
+    }
+
+    pub fn encode_auth_message(
+        &self,
+        static_key: &SecretKey,
+        remote_static_pubkey: &PublicKey,
+        buf: &mut dyn BufMut,
+    ) {
+        let mut rng = rand::thread_rng();
+        let node_id = pubkey2id(&static_key.public_key());
+
+        let message_secret_key = SecretKey::random(&mut rng);
+        let message_secret = ecdh_xchng(&message_secret_key, remote_static_pubkey);
+
+        let signature = self.sign_shared_secret(message_secret.into());
+
+        let auth = AuthMessage::new(signature, node_id, self.nonce);
+        let padding_length = rng.gen_range(100..=300);
+
+        let mut encoded_auth_msg = vec![];
+        auth.encode(&mut encoded_auth_msg);
+        encoded_auth_msg.resize(encoded_auth_msg.len() + padding_length, 0);
+
+        let ecies_data_size = 65 + 16 + 32;
+        let auth_size: u16 = (encoded_auth_msg.len() + ecies_data_size)
+            .try_into()
+            .unwrap();
+        let auth_size_bytes = auth_size.to_be_bytes();
+
+        let mut secret_keys = [0; 32];
+        kdf(&message_secret, &mut secret_keys);
+        let aes_key = &secret_keys[..16];
+        let mac_key = sha256(&secret_keys[16..]);
+
+        let iv = H128::random_using(&mut rng);
+        let mut aes_cipher = Aes128Ctr64BE::new_from_slices(aes_key, &iv.0).unwrap();
+        aes_cipher
+            .try_apply_keystream(&mut encoded_auth_msg)
+            .unwrap();
+        let encrypted_auth_msg = encoded_auth_msg;
+
+        let r_public_key = message_secret_key.public_key().to_encoded_point(false);
+        let d = sha256_hmac(&mac_key, &[&iv.0, &encrypted_auth_msg], &auth_size_bytes);
+
+        buf.put_slice(&auth_size_bytes);
+        buf.put_slice(r_public_key.as_bytes());
+        buf.put_slice(&iv.0);
+        buf.put_slice(&encrypted_auth_msg);
+        buf.put_slice(&d);
+    }
+
+    fn sign_shared_secret(&self, shared_secret: H256) -> H520 {
+        let signature_prehash = shared_secret ^ self.nonce;
+        let (signature, rid) = SigningKey::from(&self.ephemeral_key)
+            .sign_prehash_recoverable(&signature_prehash.0)
+            .unwrap();
+        let mut signature_bytes = [0; 65];
+        signature_bytes[..64].copy_from_slice(signature.to_bytes().as_slice());
+        signature_bytes[64] = rid.to_byte();
+        H520(signature_bytes)
+    }
+
+    pub fn decode_ack_message(&mut self, static_key: &SecretKey, msg: &mut [u8]) {
```

Comment:
This function could have inline comments, as well.

Reply:
Added!